tools/i18n/unescape-contents

```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env python3
+
+import argparse
+import html
+import json
+import os
+import sys
+from contextlib import contextmanager
+from pathlib import PurePath
+from typing import IO, Any, Generator, NamedTuple, Union
+
+
+class CLIArgs(NamedTuple):
+    filename: str
+    indent_level: int
+
+
+def parse_args() -> CLIArgs:
+    parser = argparse.ArgumentParser(
+        prog="unescape-contents",
+        description="Unescape Unicode and HTML entities in a JSON file",
+    )
+
+    parser.add_argument("filename", type=str)
+    parser.add_argument("--indent-level", dest="indent_level", type=int, default=2, required=False)
+    args = parser.parse_args()
+
+    return CLIArgs(args.filename, args.indent_level)
+
+
+@contextmanager
+def tempfile_for_atomic_overwrite(
+    destination: Union[str, os.PathLike[str]]
+) -> Generator[IO[Any], None, None]:
+    """
+    Creates a sibling temporary file in the same directory as `destination` and
+    yields the file-like object returned by `open()`. Upon completion of the
+    context block, the file handle is closed and the temporary file is removed,
+    either by replacing `destination` with its contents (see `os.replace()`
+    documentation) if the context succeeded, or by `os.unlink()`-ing it if the
+    context failed.
```

Comment:
At one point I documented why this was created as a sibling file instead of following `tempfile` and/or `mktemp`'s leads and creating the file in `/tmp`, but I eventually removed it as relatively unimportant. For reviewers' sakes, however: the reason is found [in the `os.replace` documentation](https://docs.python.org/3/library/os.html#os.replace):

> The operation may fail if src and dst are on different filesystems.

On my Linux system and indeed many/most modern Linuxes, `/tmp` is mounted as `tmpfs`, and `os.replace()` fails for me when moving tempfiles made by `tempfile.TemporaryFile()` and family. `tempfile.NamedTemporaryFile(dir=os.dirname(source_path))` would be a reasonable workaround if not for the fact that `TemporaryFile.__del__` unconditionally tries to unlink the file in teardown (by way of context manager or any old garbage collection run), which of course fails once the file has been `os.replace()`'d onto a new path.

Phew! Hope that helps, and as always, happy to elaborate or do some iterations on this.

Reply:
Well, oops. I went digging on this one again out of curiosity and `NamedTemporaryFile` can take a `delete` flag, so my assertion about `__del__` is only partly correct, however we still want an assurance that, by way of *either* a rename *or* an unlink, this temporary file will not survive anything but a `kill -9`, and so I tentatively still stand by `tempfile_for_atomic_overwrite` over `tempfile.NamedTemporaryFile` here.

https://github.com/python/cpython/blob/743453a5541d3a6f19046459d22a7991ecb88223/Lib/tempfile.py#L521-L546